test/connect.c

```diff
@@ -235,6 +244,86 @@ static int test_connect(struct io_uring *ring)
 	return -1;
 }
 
+static int test_connect_timeout(struct io_uring *ring)
+{
+	int connect_fd = -1, accept_fd = -1;
+	int ret, res;
+	struct sockaddr_in addr;	
+	struct io_uring_sqe *sqe;
+	int fd[1];
+
+	connect_fd = create_socket();
+	if (connect_fd == -1)
+		return -1;
+
+	accept_fd = create_socket();
+	if (accept_fd == -1)
+		goto err;
+
+	if (configure_connect(connect_fd, &addr) == -1)
+		goto err;
+	
+	ret = bind(accept_fd, (struct sockaddr*)&addr, sizeof(addr));
+	if (ret == -1) {
+		perror("bind()");
+		goto err;
+	}
+
+	ret = listen(accept_fd, 0);  // no backlog in order to block connect_fd 
+	if (ret == -1) {
+		perror("listen()");
+		goto err;
+	}
+
+	// Fill up available place in the queue
+	for (int i = 0; i < 1; ++i) {
+		fd[i] = create_socket();
+		if (configure_connect(connect_fd, &addr) == -1)
+			goto err;
+		if (connect(fd[i], &addr, sizeof(addr)) == -1) {
+			fprintf(stderr, "unable to connect %d\n", errno);
+			goto err;	
+		}
+	}
+
+	sqe = io_uring_get_sqe(ring);
+	if (!sqe) {
+		fprintf(stderr, "unable to get sqe\n");
+		goto err;
+	}
+
+	io_uring_prep_connect(sqe, connect_fd, (struct sockaddr*)&addr, sizeof(addr));
+	sqe->user_data = 1;
+	sqe->flags |= IOSQE_IO_LINK;
+
+	sqe = io_uring_get_sqe(ring);
+	if (!sqe) {
+		fprintf(stderr, "unable to get sqe\n");
+		goto err;
+	}
+	struct __kernel_timespec ts;
+	ts.tv_sec = 0;
+	ts.tv_nsec = 100000;
+	
+	io_uring_prep_link_timeout(sqe, &ts, 0);
+	ret = submit_and_wait(ring, 2, &res);
```

Comment:
It doesn't work this way, linked timeouts doesn't return an error directly but rather generate an CQE,
so you will get 2 CQEs. Tried it out a bit rewritten, as expected there are 2 CQEs, -ECANCELLED for
the connect and -ETIME for the timeout.

E.g. should be as this

```C
        ...
	ret = io_uring_submit(ring);
	if (ret != 2) {
		fprintf(stderr, "submitted %d\n", ret);
		return 1;
	}

	for (i = 0; i < 2; i++) {
		int expected;

		ret = io_uring_wait_cqe(ring, &cqe);
		if (ret) {
			fprintf(stderr, "wait_cqe=%d\n", ret);
			return 1;
		}

		expected = (cqe->user_data == 1) ? -ECANCELED : -ETIME;
		if (expected != cqe->res) {
			fprintf(stderr, "cqe %d, res %d, wanted %d\n",
				(int)cqe->user_data, cqe->res, expected);
			goto err;
		}
		io_uring_cqe_seen(ring, cqe);
	}
        ...
```

Reply:
Thanks!